File: templates/matlab/welch_psd.m
function fig = welch_psd()
    addpath(fullfile(fileparts(mfilename('fullpath')),'..','..','_utils','matlab'));
    apply_theme(); rng(0);
    fs = 1000; N = fs; t = (0:N-1)/fs;
    sig = sin(2*pi*50*t) + 0.6*sin(2*pi*120*t) + 0.3*randn(1,N);
    [Pxx, f] = pwelch(sig, 256, [], [], fs);
    fig = figure;
    semilogy(f, Pxx, 'Color', palette('cat',1), 'LineWidth', 1.5);
    xlabel('frequency (Hz)'); ylabel('PSD'); title('Welch PSD'); grid on;
end

File: templates/python/welch_psd.py
"""welch_psd: Welch 法功率谱密度（噪声信号更稳）."""
import sys, os; sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..', '_utils', 'python'))
import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import welch
from theme import apply_theme
from palette import cycle
from demo_data import gen_signal

def make_figure(sig=None, fs=None, title='Welch PSD'):
    apply_theme(fig_size=(7, 4))
    if sig is None:
        _, sig, fs = gen_signal()
    f, Pxx = welch(sig, fs=fs, nperseg=256)
    fig, ax = plt.subplots()
    ax.semilogy(f, Pxx, color=cycle(0))
    ax.set_xlabel('frequency (Hz)'); ax.set_ylabel('PSD (V²/Hz)')
    ax.set_title(title); ax.grid(True, which='both', linestyle=':', alpha=0.5)
    fig.tight_layout()
    return fig

if __name__ == '__main__':
    fig = make_figure()
    fig.savefig(__file__.replace('.py', '.png'), dpi=150)
